apps/server/src/modules/authentication/uc/account.uc.ts

```diff
@@ -86,9 +171,76 @@ export class AccountUc {
 			throw new EntityNotFoundError('User');
 		}
 		try {
-			await this.updatePassword(targetAccount, passwordNew);
+			targetAccount.password = await this.calcPasswordHash(passwordNew);
+			await this.accountRepo.update(targetAccount);
 		} catch (err) {
 			throw new EntityNotFoundError('Account');
 		}
 	}
+
+	private hasRole(user: User, roleName: string) {
+		return user.roles.getItems().some((role) => {
+			return role.name === roleName;
+		});
+	}
+
+	private isDemoUser(currentUser: User) {
+		return this.hasRole(currentUser, 'demoStudent') || this.hasRole(currentUser, 'demoTeacher');
+	}
+
+	private hasPermissionsToChangeOwnName(currentUser: User) {
+		return (
+			this.hasRole(currentUser, 'superhero') ||
+			this.hasRole(currentUser, 'teacher') ||
+			this.hasRole(currentUser, 'administrator')
+		);
+	}
+
+	private hasPermissionsToChangePassword(currentUser: User, targetUser: User) {
+		if (this.hasRole(currentUser, 'superhero')) {
+			return true;
+		}
+		if (!(currentUser.school.id === targetUser.school.id)) {
+			return false;
+		}
+		if (this.isDemoUser(currentUser)) {
+			return false;
+		}
+
+		const permissionsToCheck: string[] = [];
+		if (this.hasRole(targetUser, 'student')) {
+			permissionsToCheck.push('STUDENT_EDIT');
+		}
+		if (this.hasRole(targetUser, 'teacher')) {
+			permissionsToCheck.push('TEACHER_EDIT');
+		}
+		if (permissionsToCheck.length === 0) {
+			// target user is neither student nor teacher. Undefined what to do
+			return false;
+		}
+
+		return this.permissionService.hasUserAllSchoolPermissions(currentUser, permissionsToCheck);
+	}
+
+	private calcPasswordHash(password: string): Promise<string> {
+		return bcrypt.hash(password, 10);
+	}
+
+	private async checkPassword(enteredPassword: string, hashedAccountPassword: string) {
+		return bcrypt.compare(enteredPassword, hashedAccountPassword);
+	}
+
+	private async checkUniqueEmail(account: Account, user: User, email: string): Promise<void> {
+		const foundUsers = await this.userRepo.findByEmail(email);
```

Comment:
Better than waiting for one promise and then for the other is using `Promise.all(...)`

Reply:
done